src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2022 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jvnet.hudson.test;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.model.Slave;
+import hudson.remoting.Launcher;
+import hudson.remoting.Which;
+import hudson.slaves.DumbSlave;
+import hudson.slaves.JNLPLauncher;
+import hudson.slaves.RetentionStrategy;
+import hudson.util.StreamCopyThread;
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.tools.ant.util.JavaEnvUtils;
+import org.junit.rules.ExternalResource;
+
+/**
+ * Manages inbound agents.
+ * While these run on the local host, they are launched outside of Jenkins.
+ * @see JenkinsRule#createComputerLauncher
+ * @see JenkinsRule#createSlave()
+ */
+public final class InboundAgentRule extends ExternalResource {
+
+    private final Map<String, Process> procs = new HashMap<>();
+
+    /**
+     * Creates, attaches, and starts a new inbound agent.
+     * @param name an optional {@link Slave#getNodeName}
+     */
+    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "just for test code")
+    public Slave createAgent(@NonNull JenkinsRule r, @CheckForNull String name) throws Exception {
+        if (name == null) {
+            name = "agent" + r.jenkins.getNodes().size();
+        }
+        // TODO 2.216+ support WebSocket option
+        DumbSlave s = new DumbSlave(name, new File(r.jenkins.getRootDir(), "agent-work-dirs/" + name).getAbsolutePath(), new JNLPLauncher(true));
+        s.setNumExecutors(1); // TODO pending 2.234+
```

Comment:
A common use case I have seen in plugins is the desire to create an agent with more than one executor for complex testing scenarios, for example in `throttle-concurrent-builds`. While I recognize that this is not a use case you are intending to address in this PR, I think it would be worth thinking about how to accommodate such a use case in a future change and designing the API to support (or at least not preclude) such a future use case.

Reply:
But if I recall correctly that might be too late, since we would have already launched the agent with the wrong number of executors, and `setNumExecutors` doesn't relaunch it. And it is awkward to launch with the wrong number and then relaunch with the correct number. So I think to accommodate such a future use case the API that launches the agent would need to support customizing the number of executors prior to launch. If the API you are proposing here can be extended to support that use case, great.